### ISSUE TITLE
_message has the flash definition data instead of the final message when using custom flash elements

### ISSUE THREAD

**Issue by Zuluru** [dereuromark/cakephp-ajax#9]
In most cases, this component is working brilliantly for me, it saves so much code in my controllers! I've come across one situation where I need to do special formatting of my flash message using helpers (e.g. the Time helper to format a date and time according to user preferences), and so I'm using a custom flash element, like:

    $this->Flash->my_custom_element(null, ['params' => ['variable' => $variable]]);

And then `src/Element/Flash/my_custom_element.ctp` looks something like this:

    echo $this->element('BootstrapUI.Flash/default', [
        'message' => __('Here is the custom message about {0}',
            $this->Time->date($params['variable']->created)),
        'params' => [
            'class' => ['warning', 'alert', 'alert-dismissible', 'fade', 'in'],
            'escape' => false,
            'attributes' => ['role' => 'alert'],
        ],
    ]);

When this comes to my Ajax response handler, the _message parameter has an object with keys like so:

    element: "Flash/my_custom_element"
    key: "flash"
    message: null
    params: Object
        variable: Object
            (all the $variable data is here)

Seems that when you call `$this->Controller->request->session()->consume`, you get back an array in such cases instead of just a message, and something further needs to be done to use the element to format this before setting it into _message.

Not sure if it makes a difference that I'm using the Friends Of Cake BootstrapUI plugin to nicely format flash output, but that works fine with all my "standard" flash messages (e.g. `$this->Flash->info(__('Message goes here.'));`) Also, using this for non-Ajax requests renders the flash element correctly, so I know that much of it is okay.

**Comment by dereuromark** [comment]
If you plan on not returning the response directly, but rendering the flash messages, you should set the flash key to false, allowing the non-consumed message to be used in the rendering process of this request.

**Comment by dereuromark** [comment]
Refs https://github.com/dereuromark/cakephp-ajax/issues/8

**Comment by Zuluru** [comment]
Maybe "rendered" isn't the right word. This is an error situation with a redirect, so I need the error message returned via the Ajax response, as I won't be rendering a view. But the message needs to be formatted by the element (including using helpers) before it will be useful to my Ajax handler. Does that make sense?

**Comment by dereuromark** [comment]
yes, the question is: can the ajax view in that case manually render it instead of just returning it?
you got any ideas about that?

the alternative, would I currently do, is to have a template in JS and that is filled then. But this is not DRY and more work. So the first would be cooler, right?

**Comment by Zuluru** [comment]
Exactly right. The JS would have to know all about formatting dates and times, including the user's preferred format (e.g. 12 vs 24 hour format), and that's very not-DRY. I don't know the details about how the main framework turns those element details into HTML output, but if it's a public function, you could just call that when $message is an array.

**Comment by dereuromark** [comment]
Or you could directly call the helper and render it this way.

**Comment by Zuluru** [comment]
I've looked at the FlashHelper (both the core one and the one from BootstrapUI), and the only close-to-useful function there is render, which does more than just rendering, so it doesn't look like I could just call that and get useful results. I guess I could make my own FlashHelper by copy-paste-edit; I was hoping there might be some fairly easy thing your component could do to handle this situation for everyone that wants to use it along with custom flash elements, but maybe there's not.

**Comment by dereuromark** [comment]
I am open for ideas or PRs.

**Comment by Zuluru** [comment]
From the digging I did, it didn't seem there was an obvious, easy route, and I've got tons of other stuff to do before my release (with this not being particularly critical until close to that time), so I've put this part of my code aside for now, in hopes someone else that knows the Cake core code better will have some idea in the meantime. Or maybe a new Cake version will improve things in this area between now and then.

**Comment by dereuromark** [comment]
You could also directly use the elements and render those from the View class or via some kind of callback/closure? Maybe someone has time to look into it.

**Comment by Zuluru** [comment]
I'm looking at FlashHelper.php from both the core Cake distribution and friendsofcake bootstrap-ui. They have render functions which do a bit of session management (the same in both, except for what looks like a bug in the friendsofcake version where they delete from the session inside the loop instead of outside), then call `$this->_View->element` in a loop to build up the output, then return the output.

IF:

a) both of those were changed so that render did the session management and then called renderOne (or some such) in the loop, passing in a message to be rendered, and it was renderOne that did the various things required to generate the output

AND

b) it's somehow possible to call the registered flash helper from your component

THEN

a) friendsofcake could get rid of all their duplicated code in render and just override renderOne

AND

b) your component could call renderOne to generate the required output.

Actually, it would probably be even better if there was also a renderStack function that took the array of messages and did the loop, because the friendsofcake version does a little extra check that they might need to retain outside the loop, and then you could call the renderStack function.

Those seem like two pretty big if's to me, though...

**Comment by dereuromark** [comment]
Why not directly iterating over the flash array keys?

        $messages = [];
        foreach ($flash as $message) {
            $message = $options + $message;
            $messages[$type][] = $this->element($message['element'], $message);
        }

**Comment by Zuluru** [comment]
The problem with this is that the friendsofcake flash helper generates its output differently inside that loop. And other flash helpers might do things differently again. For this to be helper-agnostic, it seems to me that the helper itself needs to provide this, but always through a reliable API that third-parties can depend on.

**Comment by dereuromark** [comment]
You might want to make a PR for FOC then, so once this is possible we can then write the code here afterwards.

**Comment by Zuluru** [comment]
Maybe check out the FOC implementation if you haven't; part of the point is that it figures out the element to use dynamically.

And would this not need to start with a PR for Cake, and then for FOC once the Cake one is accepted?

Even if all that happened, is it possible for your component code to access the helper? Keep in mind that it has to be the helper registered through the controller, not hard-coded. I wasn't able to find the way to do that, and without it the rest seems perhaps pointless.

**Comment by dereuromark** [comment]
Yes, the view can use helpers just fine. Remember $this->Flash->... in the ctps? Same thing.
So we could always invoke FlashHelper::render() and have a complete messages HTML block.
If that suffices, the work for it is trivial (3 lines of code).
If you really need all elements to be rendered individually it will get quite complicated.

**Comment by Zuluru** [comment]
Sorry, got off on some other stuff for a while...

What you said about the view being able to use helpers got me thinking. I was assuming that the change(s) would have to be in `AjaxComponent::_respondAsAjax`, because that's where _message is set. Following your comment, I tried logging the output of `$this->Flash->render()` in `AjaxView::render`, but that was blank because `_respondAsAjax` had already consumed the flash message(s). So I commented out that part and now the log file had the flash message properly formatted. But that brings up a different (obvious in retrospect) problem, and that is that it includes HTML tags, not just the message text that my JavaScript would want. The flash element that I have for this needs to generate the HTML that way for non-Ajax responses, and I don't feel like it's a good idea to get into using different elements for Ajax vs non-Ajax.

As a result of this, it would now seem to me that my problem is probably entirely unrelated to (or at least unsolvable by) your component, and that what I need is to be able to decouple the HTML generation for the flash element from the creation of the text of the message. So, I'm going to "close and comment", and look elsewhere for a solution to how I can best call my custom date formatting helper from within a controller.
